libs/x-content/src/main/java/org/elasticsearch/common/xcontent/MediaTypeParser.java

```diff
@@ -0,0 +1,118 @@
+        for (T mediaType : acceptedMediaTypes) {
+            typeWithSubtypeToMediaType.put(mediaType.typeWithSubtype(), mediaType);
+            formatToMediaType.put(mediaType.format(), mediaType);
+        }
+    }
+
+    public MediaTypeParser(T[] acceptedMediaTypes, Map<String, T> additionalMediaTypes) {
+        this(acceptedMediaTypes);
+        for (Map.Entry<String, T> entry : additionalMediaTypes.entrySet()) {
+            String typeWithSubtype = entry.getKey();
+            T mediaType = entry.getValue();
+
+            typeWithSubtypeToMediaType.put(typeWithSubtype.toLowerCase(Locale.ROOT), mediaType);
+            formatToMediaType.put(mediaType.format(), mediaType);
+        }
+    }
+
+    public T fromMediaType(String mediaType) {
+        ParsedMediaType parsedMediaType = parseMediaType(mediaType);
+        return parsedMediaType != null ? parsedMediaType.getMediaType() : null;
+    }
+
+    public T fromFormat(String format) {
+        if (format == null) {
+            return null;
+        }
+        return formatToMediaType.get(format.toLowerCase(Locale.ROOT));
+    }
+
+    /**
+     * parsing media type that follows https://tools.ietf.org/html/rfc7231#section-3.1.1.1
+     * @param headerValue a header value from Accept or Content-Type
+     * @return a parsed media-type
+     */
+    public ParsedMediaType parseMediaType(String headerValue) {
+        if (headerValue != null) {
+            String[] split = headerValue.toLowerCase(Locale.ROOT).split(";");
+
+            String[] typeSubtype = split[0].trim().toLowerCase(Locale.ROOT)
+                .split("/");
+            if (typeSubtype.length == 2) {
+                String type = typeSubtype[0];
+                String subtype = typeSubtype[1];
+                T xContentType = typeWithSubtypeToMediaType.get(type + "/" + subtype);
+                if (xContentType != null) {
+                    Map<String, String> parameters = new HashMap<>();
+                    for (int i = 1; i < split.length; i++) {
+                        //spaces are allowed between parameters, but not between '=' sign
+                        String[] keyValueParam = split[i].trim().split("=");
+                        if (keyValueParam.length != 2 || hasSpaces(keyValueParam[0]) || hasSpaces(keyValueParam[1])) {
+                            return null;
+                        }
+                        parameters.put(keyValueParam[0].toLowerCase(Locale.ROOT), keyValueParam[1].toLowerCase(Locale.ROOT));
+                    }
+                    return new ParsedMediaType(xContentType, parameters);
+                }
+            }
+
+        }
+        return null;
+    }
+
+    private boolean hasSpaces(String s) {
+        return s.trim().equals(s) == false;
+    }
+
+    /**
+     * A media type object that contains all the information provided on a Content-Type or Accept header
+     * // TODO PG to be extended with getCompatibleAPIVersion and more
```

Comment:
can you leave these specific TODO's out. IMO TODO's left in the code base should be rare and general enough such that any future dev can understand and implement it.

Reply:
agree. removed the todo